File: mobile/modules/bluetooth-sdk/ios/Source/stt/SherpaOnnxTranscriber.swift
import Foundation
import UIKit

/**
 * SherpaOnnxTranscriber handles real-time audio transcription using Sherpa-ONNX.
 *
 * It works fully offline and processes PCM audio in real-time to provide partial and final ASR results.
 * This class runs on a background thread, processes short PCM chunks, and emits transcribed text using a delegate.
 */
class SherpaOnnxTranscriber {
    private static let TAG = "SherpaOnnxTranscriber"

    private static let SAMPLE_RATE = 16000 // Sherpa-ONNX model's required sample rate
    private static let QUEUE_CAPACITY = 100 // Max number of audio buffers to keep in queue

    private let pcmQueue = DispatchQueue(label: "com.augmentos.sherpaonnx.pcmQueue", qos: .userInteractive)
    private var pcmBuffers = [Data]()
    private var isRunning = false
    private var processingQueue: DispatchQueue?
    private var processingTask: DispatchWorkItem?

    /// The underlying Sherpa-ONNX objects
    private var recognizer: SherpaOnnxRecognizer?

    private var lastPartialResult = ""

    /// Parent context
    private weak var context: UIViewController?

    /// Session start time for relative timestamps
    private var transcriptionSessionStart: Date

    /// Dynamic model path support
    private static var customModelPath: String? {
        guard let storedPath = UserDefaults.standard.string(forKey: "STTModelPath") else {
            return nil
        }

        // Always resolve current Documents directory
        let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!

        // Extract relative subpath after "Documents/"
        // NOTE: Doing this because the application id changes between the development builds and files can't be found.
        if let range = storedPath.range(of: "/Documents/") {
            let relativePath = String(storedPath[range.upperBound...]) // e.g. "stt_models/..."
            let fixedPath = documentsURL.appendingPathComponent(relativePath).path

            Bridge.log("Reconstructed STTModelPath: \(fixedPath)")
            return fixedPath
        }

        // If nothing matched, just return as-is
        Bridge.log("STTModelPath (raw): \(storedPath)")
        return storedPath
    }

    private static func firstExistingFile(in directory: String, candidates: [String]) -> String? {
        let fileManager = FileManager.default
        for candidate in candidates {
            let path = (directory as NSString).appendingPathComponent(candidate)
            if fileManager.fileExists(atPath: path) {
                return path
            }
        }
        return nil
    }

    /**
     * Constructor that accepts a UIViewController to load model assets.
     */
    init(context: UIViewController) {
        self.context = context
        transcriptionSessionStart = Date()
    }

    deinit {
        shutdown()
    }

    /**
     * Initialize the Sherpa-ONNX recognizer.
     * Loads models and configuration, sets up processing thread.
     */
    func initialize() {
        do {
            var tokensPath: String
            var modelType = "unknown"
            let fileManager = FileManager.default

            // Check if we have a custom model path set
            if let customPath = SherpaOnnxTranscriber.customModelPath {
                // Detect model type based on available files
                let ctcModelPath = Self.firstExistingFile(
                    in: customPath,
                    candidates: ["model.int8.onnx", "model.onnx"]
                )
                let transducerEncoderPath = Self.firstExistingFile(
                    in: customPath,
                    candidates: ["encoder.int8.onnx", "encoder.onnx"]
                )

                tokensPath = (customPath as NSString).appendingPathComponent("tokens.txt")

                // Verify tokens file exists
                guard fileManager.fileExists(atPath: tokensPath) else {
                    throw NSError(domain: "SherpaOnnxTranscriber", code: 1, userInfo: [
                        NSLocalizedDescriptionKey: "tokens.txt not found at path: \(customPath)",
                    ])
                }

                if let ctcModelPath {
                    // CTC model detected
                    modelType = "ctc"
                    Bridge.log("Detected CTC model at \(customPath)")

                    // Create CTC model config using Zipformer2Ctc
                    var nemoCtc = sherpaOnnxOnlineNemoCtcModelConfig(
                        model: ctcModelPath
                    )

                    // Create model config with CTC
                    var modelConfig = sherpaOnnxOnlineModelConfig(
                        tokens: tokensPath,
                        numThreads: 1,
                        nemoCtc: nemoCtc
                    )

                    // Configure recognizer
                    var featureConfig = sherpaOnnxFeatureConfig()

                    var config = sherpaOnnxOnlineRecognizerConfig(
                        featConfig: featureConfig,
                        modelConfig: modelConfig,
                        enableEndpoint: true,
                        rule1MinTrailingSilence: 1.2,
                        rule2MinTrailingSilence: 0.8,
                        rule3MinUtteranceLength: 10.0
                    )

                    // Create recognizer with the wrapper
                    recognizer = SherpaOnnxRecognizer(config: &config)

                } else if let transducerEncoderPath {
                    // Transducer model detected
                    modelType = "transducer"
                    Bridge.log("Detected transducer model at \(customPath)")

                    let decoderPath = Self.firstExistingFile(
                        in: customPath,
                        candidates: ["decoder.int8.onnx", "decoder.onnx"]
                    )
                    let joinerPath = Self.firstExistingFile(
                        in: customPath,
                        candidates: ["joiner.int8.onnx", "joiner.onnx"]
                    )

                    // Verify all transducer files exist
                    guard let decoderPath,
                          let joinerPath
                    else {
                        throw NSError(domain: "SherpaOnnxTranscriber", code: 1, userInfo: [
                            NSLocalizedDescriptionKey: "Transducer model files incomplete at path: \(customPath)",
                        ])
                    }

                    // Create Sherpa-ONNX transducer model config
                    var transducer = sherpaOnnxOnlineTransducerModelConfig(
                        encoder: transducerEncoderPath,
                        decoder: decoderPath,
                        joiner: joinerPath
                    )

                    // Create model config
                    var modelConfig = sherpaOnnxOnlineModelConfig(
                        tokens: tokensPath,
                        transducer: transducer,
                        numThreads: 1
                    )

                    // Configure recognizer
                    var featureConfig = sherpaOnnxFeatureConfig()

                    var config = sherpaOnnxOnlineRecognizerConfig(
                        featConfig: featureConfig,
                        modelConfig: modelConfig,
                        enableEndpoint: true,
                        rule1MinTrailingSilence: 1.2,
                        rule2MinTrailingSilence: 0.8,
                        rule3MinUtteranceLength: 10.0
                    )

                    // Create recognizer with the wrapper
                    recognizer = SherpaOnnxRecognizer(config: &config)

                } else {
                    throw NSError(domain: "SherpaOnnxTranscriber", code: 1, userInfo: [
                        NSLocalizedDescriptionKey: "No valid model files found at path: \(customPath)",
                    ])
                }
            } else {
                Bridge.log("No Sherpa ONNX model available. Transcription will be disabled.")
                Bridge.log("Please download a model using the model downloader in settings.")
                recognizer = nil
                isRunning = false
                return
            }

            if recognizer == nil {
                throw NSError(domain: "SherpaOnnxTranscriber", code: 2, userInfo: [NSLocalizedDescriptionKey: "Failed to create recognizer"])
            }

            startProcessingTask()
            isRunning = true

            Bridge.log("Sherpa-ONNX ASR initialized successfully with \(modelType) model")

        } catch {
            Bridge.log("Failed to initialize Sherpa-ONNX: \(error.localizedDescription)")
        }
    }

    /**
     * Handle transcription results - send only to delegate
     */
    private func handleTranscriptionResult(text: String, isFinal: Bool) {
        // Forward to delegate if set
        DispatchQueue.main.async { [weak self] in
            if isFinal {
                STTTools.didReceiveFinalTranscription(text)
            } else {
                STTTools.didReceivePartialTranscription(text)
            }
        }
    }

    /**
     * Feed PCM audio data (16-bit little endian) into the transcriber.
     * This method should be called continuously with short chunks (e.g., 100-300ms).
     *
     * Audio is queued directly; microphone VAD gating is not applied in the SDK.
     */
    func acceptAudio(pcm16le: Data) {
        guard isRunning else {
            return
        }

        queueAudioData(pcm16le)
    }

    private func queueAudioData(_ pcm16le: Data) {
        pcmQueue.async { [weak self] in
            guard let self = self else { return }

            let queueSizeBefore = self.pcmBuffers.count
            self.pcmBuffers.append(pcm16le)

            // Keep queue size manageable
            if self.pcmBuffers.count > Self.QUEUE_CAPACITY {
                let removedBuffer = self.pcmBuffers.removeFirst()
                Bridge.log("⚠️ Audio queue overflow - dropped buffer of \(removedBuffer.count) bytes")
            }
        }
    }

    /**
     * Start a background task to continuously consume audio and decode using Sherpa.
     */
    private func startProcessingTask() {
        Bridge.log("🚀 Starting Sherpa-ONNX processing task...")

        processingQueue = DispatchQueue(label: "com.augmentos.sherpaonnx.processor", qos: .userInitiated)

        let workItem = DispatchWorkItem { [weak self] in
            self?.runLoop()
        }

        processingTask = workItem
        processingQueue?.async(execute: workItem)
    }

    /**
     * Main processing loop that handles transcription in real-time.
     * Pulls audio from queue, feeds into Sherpa, emits partial/final results.
     */
    private func runLoop() {
        Bridge.log("🔄 Sherpa-ONNX processing loop started")

        while isRunning {
            // Pull data from queue
            var audioData: Data?

            pcmQueue.sync {
                if !self.pcmBuffers.isEmpty {
                    audioData = self.pcmBuffers.removeFirst()
                }
            }

            if let data = audioData {
                // Synchronize access to recognizer to prevent race conditions
                objc_sync_enter(self)
                defer { objc_sync_exit(self) }

                guard let recognizer = recognizer else {
                    Bridge.log("⚠️ Recognizer not available, skipping audio chunk")
                    continue
                }

                do {
                    // Convert PCM to float [-1.0, 1.0]
                    let floatBuf = toFloatArray(from: data)

                    // Pass audio data to the Sherpa-ONNX stream
                    recognizer.acceptWaveform(samples: floatBuf, sampleRate: Self.SAMPLE_RATE)

                    // Decode continuously while model is ready
                    var decodeCount = 0
                    while recognizer.isReady() {
                        recognizer.decode()
                        decodeCount += 1
                    }

                    // If utterance endpoint detected
                    if recognizer.isEndpoint() {
                        let result = recognizer.getResult()
                        let finalText = result.text.trimmingCharacters(in: .whitespacesAndNewlines)

                        if !finalText.isEmpty {
                            handleTranscriptionResult(text: finalText, isFinal: true)
                        }

                        recognizer.reset() // Start new utterance
                        lastPartialResult = ""
                    } else {
                        // Emit partial results if changed
                        let result = recognizer.getResult()
                        let partial = result.text.trimmingCharacters(in: .whitespacesAndNewlines)

                        if partial != lastPartialResult, !partial.isEmpty {
                            handleTranscriptionResult(text: partial, isFinal: false)
                            lastPartialResult = partial
                        }
                    }
                } catch {
                    Bridge.log("❌ Error processing audio: \(error.localizedDescription)")
                }
            } else {
                // Sleep briefly to avoid tight CPU loop if no audio is available
                Thread.sleep(forTimeInterval: 0.01)
            }
        }

        Bridge.log("ASR processing thread stopped")
    }

    /**
     * Convert 16-bit PCM byte data (little-endian) to float array [-1.0, 1.0].
     */
    private func toFloatArray(from pcmData: Data) -> [Float] {
        let count = pcmData.count / 2
        var samples = [Float](repeating: 0, count: count)

        pcmData.withUnsafeBytes { (bufferPointer: UnsafeRawBufferPointer) in
            if let address = bufferPointer.baseAddress {
                let int16Pointer = address.bindMemory(to: Int16.self, capacity: count)

                for i in 0 ..< count {
                    // Convert from little-endian if needed
                    var sample = int16Pointer[i]
                    if CFByteOrderGetCurrent() == CFByteOrder(CFByteOrderBigEndian.rawValue) {
                        sample = Int16(littleEndian: sample)
                    }
                    samples[i] = Float(sample) / 32768.0
                }
            }
        }

        return samples
    }

    /**
     * Stop transcription processing.
     * This shuts down the processing thread and releases Sherpa-ONNX resources.
     */
    func shutdown() {
        Bridge.log("🛑 Shutting down SherpaOnnxTranscriber...")

        isRunning = false
        processingTask?.cancel()

        // Synchronize access to recognizer during shutdown
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }

        // The recognizer will be automatically cleaned up by ARC when set to nil
        if recognizer != nil {
            Bridge.log("🧹 Cleaning up Sherpa-ONNX recognizer")
            recognizer = nil
        }

        // Clear any remaining audio buffers
        pcmQueue.sync {
            let remainingBuffers = self.pcmBuffers.count
            if remainingBuffers > 0 {
                Bridge.log("🗑️ Clearing \(remainingBuffers) remaining audio buffers")
            }
            self.pcmBuffers.removeAll()
        }

        Bridge.log("✅ SherpaOnnxTranscriber shutdown complete")
    }

    /**
     * Restarts the transcriber after a model change.
     * Shuts down existing resources, clears buffers, and reinitializes the recognizer.
     */
    func restart() {
        Bridge.log("♻️ Restarting SherpaOnnxTranscriber...")
        shutdown()
        initialize()
    }
}

File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/stt/SherpaOnnxTranscriber.kt
package com.mentra.bluetoothsdk.stt

import android.content.Context
import android.content.SharedPreferences
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.k2fsa.sherpa.onnx.*
import com.mentra.bluetoothsdk.Bridge
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.BlockingQueue
import java.util.concurrent.atomic.AtomicBoolean

/**
 * SherpaOnnxTranscriber handles real-time audio transcription using Sherpa-ONNX.
 *
 * It works fully offline and processes PCM audio in real-time to provide partial and final ASR
 * results. This class runs on a background thread, processes short PCM chunks, and emits
 * transcribed text using a listener.
 *
 * Ported from iOS SherpaOnnxTranscriber.swift to match iOS functionality 1:1
 */
class SherpaOnnxTranscriber(private val context: Context) {
    companion object {
        private const val TAG = "SherpaOnnxTranscriber"
        private const val SAMPLE_RATE = 16000 // Sherpa-ONNX model's required sample rate
        private const val QUEUE_CAPACITY = 100 // Max number of audio buffers to keep in queue
        private const val PREFS_NAME = "MentraPrefs"
        private const val KEY_STT_MODEL_PATH = "STTModelPath"
        private const val KEY_STT_MODEL_LANGUAGE = "STTModelLanguageCode"
    }

    /** Interface to receive transcription results from Sherpa-ONNX */
    interface TranscriptListener {
        /** Called with live partial transcription (not final yet) */
        fun onPartialResult(text: String, language: String)

        /** Called when an utterance ends and final text is available */
        fun onFinalResult(text: String, language: String)
    }

    private val pcmQueue: BlockingQueue<ByteArray> = ArrayBlockingQueue(QUEUE_CAPACITY)
    private val running = AtomicBoolean(false)
    private val mainHandler = Handler(Looper.getMainLooper())

    private var workerThread: Thread? = null
    private var recognizer: OnlineRecognizer? = null
    private var stream: OnlineStream? = null

    private var lastPartialResult = ""
    @Volatile private var transcriptListener: TranscriptListener? = null

    private val restartLock = Any()
    private var restartRunning = false

    private val prefs: SharedPreferences by lazy {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    /**
     * Get the custom model path from SharedPreferences Resolves relative paths from Documents
     * directory
     */
    private fun getCustomModelPath(): String? {
        val storedPath = prefs.getString(KEY_STT_MODEL_PATH, null) ?: return null

        // Always resolve to absolute path if it contains "Documents/"
        // This handles app ID changes between dev builds
        if (storedPath.contains("/Documents/")) {
            val documentsDir =
                    context.filesDir // In Android, this is the app's private files directory
            val relativePath = storedPath.substringAfter("/Documents/")
            val fixedPath = File(documentsDir, relativePath).absolutePath
            Bridge.log("Reconstructed STTModelPath: $fixedPath")
            return fixedPath
        }

        Bridge.log("STTModelPath (raw): $storedPath")
        return storedPath
    }

    /** Get the model language code from SharedPreferences */
    private fun getModelLanguage(): String {
        return prefs.getString(KEY_STT_MODEL_LANGUAGE, "en-US") ?: "en-US"
    }

    /**
     * Initialize the Sherpa-ONNX recognizer. Loads models and configuration, sets up processing
     * thread.
     */
    fun initialize() {
        try {
            val modelPath = getCustomModelPath()

            if (modelPath == null) {
                Bridge.log("No Sherpa ONNX model available. Transcription will be disabled.")
                Bridge.log("Please download a model using the model downloader in settings.")
                recognizer = null
                stream = null
                return
            }

            val modelDir = File(modelPath)
            if (!modelDir.exists() || !modelDir.isDirectory) {
                Bridge.log("Model path does not exist or is not a directory: $modelPath")
                recognizer = null
                stream = null
                return
            }

            val tokensPath = File(modelDir, "tokens.txt").absolutePath
            val tokensFile = File(tokensPath)

            if (!tokensFile.exists()) {
                throw IllegalStateException("tokens.txt not found at path: $modelPath")
            }

            val modelType: String
            val modelConfig = OnlineModelConfig()

            // Detect model type based on available files
            val ctcModelFile = findReadableFile(modelDir, listOf("model.int8.onnx", "model.onnx"))
            val transducerEncoderFile =
                    findReadableFile(modelDir, listOf("encoder.onnx", "encoder.int8.onnx"))

            when {
                ctcModelFile.exists() -> {
                    // CTC model detected
                    modelType = "ctc"
                    Bridge.log("Detected CTC model at $modelPath")

                    val nemoCtc = OnlineNeMoCtcModelConfig()
                    nemoCtc.model = ctcModelFile.absolutePath

                    modelConfig.tokens = tokensPath
                    modelConfig.numThreads = 1
                    modelConfig.neMoCtc = nemoCtc
                }
                transducerEncoderFile.exists() -> {
                    // Transducer model detected
                    modelType = "transducer"
                    Bridge.log("Detected transducer model at $modelPath")

                    val decoderFile =
                            findReadableFile(modelDir, listOf("decoder.onnx", "decoder.int8.onnx"))
                    val joinerFile =
                            findReadableFile(modelDir, listOf("joiner.onnx", "joiner.int8.onnx"))

                    if (!decoderFile.exists() || !joinerFile.exists()) {
                        throw IllegalStateException(
                                "Transducer model files incomplete at path: $modelPath"
                        )
                    }

                    val transducer = OnlineTransducerModelConfig()
                    transducer.encoder = transducerEncoderFile.absolutePath
                    transducer.decoder = decoderFile.absolutePath
                    transducer.joiner = joinerFile.absolutePath

                    modelConfig.tokens = tokensPath
                    modelConfig.transducer = transducer
                    modelConfig.numThreads = 1
                }
                else -> {
                    throw IllegalStateException("No valid model files found at path: $modelPath")
                }
            }

            // Configure recognizer
            val config = OnlineRecognizerConfig()
            config.modelConfig = modelConfig
            config.decodingMethod = "greedy_search"
            config.enableEndpoint = true

            // Create recognizer (pass null for AssetManager since we're using file paths)
            recognizer =
                    try {
                        OnlineRecognizer(null, config)
                    } catch (e: Exception) {
                        Log.e(
                                TAG,
                                "Failed to create OnlineRecognizer - model file may be corrupted or incomplete",
                                e
                        )
                        null
                    }

            if (recognizer == null) {
                throw IllegalStateException("Failed to create recognizer")
            }

            stream = recognizer?.createStream("")

            startProcessingThread()
            running.set(true)

            Bridge.log("Sherpa-ONNX ASR initialized successfully with $modelType model")
        } catch (e: Exception) {
            Bridge.log("Failed to initialize Sherpa-ONNX: ${e.message}")
            Log.e(TAG, "Failed to initialize Sherpa-ONNX", e)

            // Clean up any partially initialized resources
            stream?.let {
                try {
                    it.release()
                } catch (releaseEx: Exception) {
                    Log.e(TAG, "Error releasing stream after initialization failure", releaseEx)
                }
            }
            stream = null

            recognizer?.let {
                try {
                    it.release()
                } catch (releaseEx: Exception) {
                    Log.e(TAG, "Error releasing recognizer after initialization failure", releaseEx)
                }
            }
            recognizer = null

            running.set(false)
        }
    }

    private fun findReadableFile(modelDir: File, candidates: List<String>): File {
        return candidates.map { File(modelDir, it) }.firstOrNull { file ->
            file.exists() && file.canRead() && file.length() > 0
        } ?: File(modelDir, candidates.first())
    }

    /**
     * Feed PCM audio data (16-bit little endian) into the transcriber. This method should be called
     * continuously with short chunks (e.g., 100-300ms).
     *
     * Audio is queued directly; microphone VAD gating is not applied in the SDK.
     */
    fun acceptAudio(pcm16le: ByteArray) {
        if (!running.get()) {
            // Bridge.log("⚠️ Ignoring audio - transcriber not running")
            return
        }

        if (!pcmQueue.offer(pcm16le.copyOf())) {
            // Queue is full, drop oldest and try again
            pcmQueue.poll()
            pcmQueue.offer(pcm16le.copyOf())
            Bridge.log("⚠️ Audio queue overflow - dropped buffer")
        }
    }

    /** Start a background thread to continuously consume audio and decode using Sherpa */
    private fun startProcessingThread() {
        Bridge.log("🚀 Starting Sherpa-ONNX processing thread...")

        workerThread =
                Thread({ runLoop() }, "SherpaOnnxProcessor").apply {
                    isDaemon = true
                    start()
                }
    }

    /**
     * Main processing loop that handles transcription in real-time. Pulls audio from queue, feeds
     * into Sherpa, emits partial/final results.
     */
    private fun runLoop() {
        Bridge.log("🔄 Sherpa-ONNX processing loop started")

        while (running.get()) {
            try {
                val currentRecognizer = recognizer
                val currentStream = stream

                if (currentRecognizer == null || currentStream == null) {
                    Bridge.log("⚠️ Recognizer or stream not available, skipping audio chunk")
                    Thread.sleep(100)
                    continue
                }

                // Pull data from queue with timeout
                val audioData = pcmQueue.poll(50, java.util.concurrent.TimeUnit.MILLISECONDS)

                if (audioData != null) {
                    // Convert PCM to float [-1.0, 1.0]
                    val floatBuf = toFloatArray(audioData)

                    // Pass audio data to the Sherpa-ONNX stream
                    currentStream.acceptWaveform(floatBuf, SAMPLE_RATE)

                    // Decode continuously while model is ready
                    var decodeCount = 0
                    while (currentRecognizer.isReady(currentStream)) {
                        currentRecognizer.decode(currentStream)
                        decodeCount++
                    }

                    // If utterance endpoint detected
                    if (currentRecognizer.isEndpoint(currentStream)) {
                        val result = currentRecognizer.getResult(currentStream)
                        val finalText = result.text.trim()

                        if (finalText.isNotEmpty()) {
                            handleTranscriptionResult(finalText, isFinal = true)
                        }

                        currentRecognizer.reset(currentStream) // Start new utterance
                        lastPartialResult = ""
                    } else {
                        // Emit partial results if changed
                        val result = currentRecognizer.getResult(currentStream)
                        val partial = result.text.trim()

                        if (partial != lastPartialResult && partial.isNotEmpty()) {
                            handleTranscriptionResult(partial, isFinal = false)
                            lastPartialResult = partial
                        }
                    }
                }
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
                Bridge.log("Processing thread interrupted")
                break
            } catch (e: Exception) {
                Log.e(TAG, "Error processing audio", e)
                // Attempt stream reset to recover
                try {
                    recognizer?.let { rec -> stream?.let { str -> rec.reset(str) } }
                } catch (resetEx: Exception) {
                    Log.e(TAG, "Failed to reset stream after error", resetEx)
                }
            }
        }

        Bridge.log("ASR processing thread stopped")
    }

    /** Convert 16-bit PCM byte data (little-endian) to float array [-1.0, 1.0] */
    private fun toFloatArray(pcmData: ByteArray): FloatArray {
        val count = pcmData.size / 2
        val samples = FloatArray(count)

        val buffer = ByteBuffer.wrap(pcmData).order(ByteOrder.LITTLE_ENDIAN)

        for (i in 0 until count) {
            samples[i] = buffer.short / 32768.0f
        }

        return samples
    }

    /** Handle transcription results - send to listener on main thread */
    private fun handleTranscriptionResult(text: String, isFinal: Boolean) {
        val language = getModelLanguage()

        // Forward to listener on main thread
        mainHandler.post {
            transcriptListener?.let { listener ->
                if (isFinal) {
                    listener.onFinalResult(text, language)
                } else {
                    listener.onPartialResult(text, language)
                }
            }
        }
    }

    /**
     * Stop transcription processing. This shuts down the processing thread and releases Sherpa-ONNX
     * resources.
     */
    fun shutdown() {
        Bridge.log("🛑 Shutting down SherpaOnnxTranscriber...")

        running.set(false)

        workerThread?.let { thread ->
            thread.interrupt()
            try {
                thread.join(500)
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
            }
        }
        workerThread = null

        try {
            stream?.let {
                Bridge.log("🧹 Cleaning up Sherpa-ONNX stream")
                it.release()
            }
            stream = null

            recognizer?.let {
                Bridge.log("🧹 Cleaning up Sherpa-ONNX recognizer")
                it.release()
            }
            recognizer = null
        } catch (e: Exception) {
            Log.e(TAG, "Error releasing Sherpa resources", e)
        }

        // Clear any remaining audio buffers
        val remainingBuffers = pcmQueue.size
        if (remainingBuffers > 0) {
            Bridge.log("🗑️ Clearing $remainingBuffers remaining audio buffers")
        }
        pcmQueue.clear()

        Bridge.log("✅ SherpaOnnxTranscriber shutdown complete")
    }

    /**
     * Restarts the transcriber after a model change. Shuts down existing resources, clears buffers,
     * and reinitializes the recognizer.
     */
    fun restart() {
        synchronized(restartLock) {
            if (restartRunning) {
                Bridge.log("Restart already in progress, skipping")
                return
            }
            restartRunning = true
        }

        try {
            Bridge.log("♻️ Restarting SherpaOnnxTranscriber...")
            shutdown()
            // Small delay to ensure cleanup completes
            Thread.sleep(100)
            initialize()
        } catch (e: Exception) {
            Log.e(TAG, "Error during transcriber restart", e)
        } finally {
            synchronized(restartLock) { restartRunning = false }
        }
    }

    /** Register a listener to receive partial and final transcription updates */
    fun setTranscriptListener(listener: TranscriptListener?) {
        transcriptListener = listener
    }

    /** Check if the transcriber was successfully initialized */
    fun isInitialized(): Boolean {
        return recognizer != null && stream != null
    }

    /** Handle microphone state changes */
    fun microphoneStateChanged(state: Boolean) {
        if (!state) {
            // Microphone turned off - clear queue and reset stream
            pcmQueue.clear()

            recognizer?.let { rec ->
                stream?.let { str ->
                    try {
                        rec.reset(str)
                        lastPartialResult = ""
                        Log.d(TAG, "Microphone off — stream reset")
                    } catch (e: Exception) {
                        Log.e(TAG, "Error resetting stream on mic off", e)
                    }
                }
            }
        } else {
            Log.d(TAG, "Microphone on")
        }
    }
}
